tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncLineDebuggerInvoker.cs

```diff
@@ -0,0 +1,231 @@
+            if (!ProbeRateLimiter.Instance.Sample(probeId))
+            {
+                return CreateInvalidatedAsyncLineDebuggerState();
+            }
+
+            var kickoffInfo = AsyncHelper.GetAsyncKickoffMethodInfo(instance);
+            if (kickoffInfo.KickoffParentObject == null && kickoffInfo.KickoffMethod.IsStatic == false)
+            {
+                Log.Error($"{nameof(BeginLine)}: hoisted 'this' has not found. {kickoffInfo.KickoffParentType.Name}.{kickoffInfo.KickoffMethod.Name}");
+                return CreateInvalidatedAsyncLineDebuggerState();
+            }
+
+            if (!MethodMetadataProvider.TryCreateIfNotExists(instance, methodMetadataIndex, in methodHandle, in typeHandle, kickoffInfo))
+            {
+                Log.Warning($"BeginMethod_StartMarker: Failed to receive the InstrumentedMethodInfo associated with the executing method. type = {typeof(TTarget)}, instance type name = {instance?.GetType().Name}, methodMetadaId = {methodMetadataIndex}");
+                return CreateInvalidatedAsyncLineDebuggerState();
+            }
+
+            var state = new AsyncLineDebuggerState(probeId, scope: default, DateTimeOffset.UtcNow, methodMetadataIndex, lineNumber, probeFilePath, instance, kickoffInfo.KickoffParentObject);
+            state.SnapshotCreator.StartDebugger();
+            state.SnapshotCreator.StartSnapshot();
+            state.SnapshotCreator.StartCaptures();
+            state.SnapshotCreator.StartLines(lineNumber);
+
+            return state;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void CollectArgs(ref AsyncLineDebuggerState asyncState)
+        {
+            // capture hoisted arguments
+            var kickOffMethodArguments = asyncState.MethodMetadataInfo.AsyncMethodHoistedArguments;
+            for (var index = 0; index < kickOffMethodArguments.Length; index++)
+            {
+                ref var argument = ref kickOffMethodArguments[index];
+                if (argument == default)
+                {
+                    continue;
+                }
+
+                var argumentValue = argument.GetValue(asyncState.MoveNextInvocationTarget);
+                LogArg(ref argumentValue, argument.FieldType, argument.Name, index, ref asyncState);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void CollectLocals(ref AsyncLineDebuggerState asyncState)
+        {
+            // MethodMetadataInfo saves locals from MoveNext localVarSig,
+            // this isn't enough in async scenario because we need to extract more locals the may hoisted in the builder object
+            // and we need to subtract some locals that exist in the localVarSig but they are not belongs to the kickoff method
+            // For know we capturing here all locals the are hoisted (except known generated locals)
+            // and we capturing in LogLocal the locals form localVarSig
```

Comment:
```suggestion
            // In the async scenario MethodMetadataInfo stores locals from MoveNext's localVarSig,
            // which isn't enough because we need to extract more locals that may be hoisted in the builder object
            // and we need to remove some locals that exist in the localVarSig but are just part of the async machinery and do not represent actual variables in the user's code.
```